### PR TITLE
Support to parse numbers in text-based input formats

### PR DESCRIPTION
**Pull request by abhishekrb19**
Text-based input formats like `csv` and `tsv` currently parse inputs only as strings, following the `RFC4180Parser` spec).
To workaround this, the web-console and other tools need to further inspect the sample data returned to sample data returned by the Druid sampler API to parse them as numbers. See [here](https://github.com/apache/druid/blob/master/web-console/src/druid-models/ingestion-spec/ingestion-spec.tsx) for the relevant web-console code.

Changes:
- Introduce a new optional config `shouldParseNumbers` for the `csv` and `tsv` input formats.
- If enabled, the input format reader will parse numeric values from the input string for these text-based input formats.
- By default, this configuration is false for backwards compatibility
- Note that when enabled, this can have a small performance hit as the input needs to be parsed to determined the numeric type.

Key classes to review:
- `ParserUtils`
- `CsvInputFormat`
- `CsvParser`
- `DelimitedInputFormat`
- `DelimitedValueReader`

#### Release note:

Introduce a new optional config, `tryParseNumbers`, for the `csv` and `tsv` input formats.If enabled, any numbers present in the input will be parsed in the following manner --  `long` data type for integer types and `double` for floating-point numbers. By default, this configuration is set to false, so numeric strings will be treated as strings.

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: docs/ingestion/data-formats.md [hunk 1]
```diff
@@ -125,6 +125,8 @@ Configure the CSV `inputFormat` to load CSV data as follows:
 | columns | JSON array | Specifies the columns of the data. The columns should be in the same order with the columns of your data. | yes if `findColumnsFromHeader` is false or missing |
 | findColumnsFromHeader | Boolean | If this is set, the task will find the column names from the header row. Note that `skipHeaderRows` will be applied before finding column names from the header. For example, if you set `skipHeaderRows` to 2 and `findColumnsFromHeader` to true, the task will skip the first two lines and then extract column information from the third line. `columns` will be ignored if this is set to true. | no (default = false if `columns` is set; otherwise null) |
 | skipHeaderRows | Integer | If this is set, the task will skip the first `skipHeaderRows` rows. | no (default = 0) |
+| shouldParseNumbers| Boolean| If this is set, the task will attempt to parse numeric strings into long or double data type, in that order. If the value cannot be parsed as a number, it is retained as a string. | no (default = false) |
```
Comment (clintropolis) [inline docs/ingestion/data-formats.md:128]: i wonder if this should be `tryParseNumbers`? this is fine too though...

File: processing/src/main/java/org/apache/druid/java/util/common/parsers/ParserUtils.java [hunk 2]
```diff
@@ -52,22 +55,61 @@ public class ParserUtils
...
 
+  /**
+   * Attempts to parse the input string into a numeric value, if applicable. If the input is a number, the method first
+   * tries to parse the input number as a {@code Long}. If parsing as a {@code Long} fails, it then attempts to parse
+   * the input number as a {@code Double}. For all other scenarios, the input is returned as-is as a {@code String} type.
+   */
+  @Nullable
+  private static Object tryParseStringAsNumber(@Nullable final String input)
+  {
+    if (!NumberUtils.isNumber(input)) {
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/java/util/common/parsers/ParserUtils.java:97]: i wonder if this is worth looping over the string an extra time before we do try parse attempts, or if we should just start with trying to parse it as a long. I guess having this function call saves the double tryParse which uses a regex pattern.

File: processing/src/main/java/org/apache/druid/java/util/common/parsers/ParserUtils.java [hunk 3]
```diff
@@ -52,22 +55,61 @@ public class ParserUtils
...
   }
 
-  public static Function<String, Object> getMultiValueFunction(
+  /**
+   * @return a function that processes a given string input by splitting it into multiple values
+   * using the {@code listSplitter} if thge {@code list delimiter} is present in the input. If {@code shouldParseNumbers}
+   * is enabled, the function will also try to parse any numeric values present in the input -- integers as {@code Long}
+   * and floating-point numbers as {@code Double}.
+   */
+  public static Function<String, Object> getMultiValueAndParseNumbersFunction(
```
[the remark this answers is not in the document]
Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/java/util/common/parsers/ParserUtils.java]: Renamed to `getTransformationFunction`

File: processing/src/main/java/org/apache/druid/java/util/common/parsers/ParserUtils.java (hunk 2 above)
Comment (abhishekrb19) [reply processing/src/main/java/org/apache/druid/java/util/common/parsers/ParserUtils.java:98]: Yeah, I considered something like that. However, it adds an additional overhead of regex that you note for string inputs, so I kept the current approach, which is optimized for strings

**Review by clintropolis — APPROVED** [review]

**Pull request merged**